test/abstract_syntax_tree.coffee

```diff
@@ -10,207 +10,595 @@ getExpressions = (code) ->
 
 getExpression = (code) -> getExpressions(code)[0]
 
+# Recursively compare all values of enumerable properties of `expected` with
+# those of `actual`. Use `looseArray` helper function to skip array length
+# comparison.
+
+deepStrictEqualExpectedProps = (actual, expected) ->
+  white = (text, values...) -> (text[i] + "#{reset}#{v}#{red}" for v, i in values).join('') + text[i]
+  eq actual.length, expected.length if expected instanceof Array and not expected.loose
+  for k , v of expected
+    if 'object' is typeof v
+      fail white"`actual` misses #{k} property." unless k of actual
+      deepStrictEqualExpectedProps actual[k], v
+    else
+      eq actual[k], v, white"Property #{k}: expected #{actual[k]} to equal #{v}"
+  actual
+
+# Flag array for loose comparision. See above code/comment.
+looseArray = (arr) ->
+  Object.defineProperty arr, 'loose',
+    value: yes
+    enumerable: no
+  arr
+
+testExpressions = (code, expected) ->
+  ast = getExpressions code
+  return console.log ast unless expected?
+  deepStrictEqualExpectedProps ast, expected
+
+testExpression = (code, expected) ->
+  ast = getExpression code
+  return console.log ast unless expected?
+  deepStrictEqualExpectedProps ast, expected
+
+
+test 'Confirm functionality of `deepStrictEqualExpectedProps`', ->
+  actual =
+    name: 'Name'
+    a:
+      b: 1
+      c: 2
+    x: [1, 2, 3]
+
+  check = (message, test, expected) ->
+    test (-> deepStrictEqualExpectedProps actual, expected), message
+
+  check 'Expected prop does not match', throws,
+    name: '"Name"'
+
+  check 'Array length mismatch', throws,
+    x: [1, 2]
+
+  check 'Skip array length check', doesNotThrow,
+    x: looseArray [
+      1
+      2
+    ]
+
+  check 'Array length matches', doesNotThrow,
+    x: [1, 2, 3]
+
+  check 'Array prop mismatch', throws,
+    x: [3, 2, 1]
+
+  check 'Partial object comparison', doesNotThrow,
+    a: c: 2
```

Comment:
This line looks confusing. Is it:

```coffee
a:
  c: 2
```
?

Reply:
Yes it is. To me those aren't confusing at all. It's just coffeescript.